ibm/resource_ibm_securitygroup_rule.go

```diff
@@ -0,0 +1,253 @@
+		sgrule.PortRangeMax = sl.Int(d.Get("port_range_max").(int))
+
+	}
+
+	if d.Get("protocol").(string) != "" {
+		sgrule.Protocol = sl.String(d.Get("protocol").(string))
+	}
+
+	// if only one of min/max is provided, set the other one to the provided
+	if sgrule.PortRangeMin != nil && sgrule.PortRangeMax == nil {
+		sgrule.PortRangeMax = sgrule.PortRangeMin
+	}
+	if sgrule.PortRangeMax != nil && sgrule.PortRangeMin == nil {
+		sgrule.PortRangeMin = sgrule.PortRangeMax
+	}
+
+	if sgrule.PortRangeMin != nil {
+		if *sgrule.PortRangeMax < *sgrule.PortRangeMin {
+			return fmt.Errorf("port range max should be greater or equal to min")
+		}
+
+		if sgrule.Protocol == nil {
+			return fmt.Errorf("Protocol has to be provided if port range is used.")
+		}
+	}
+
+	matchingrule, err := findMatchingRule(securitygroupid, &sgrule, service)
+	if err != nil {
+		return err
+	}
+
+	if matchingrule != nil {
+		log.Printf("[INFO] rule exists")
+		d.SetId(fmt.Sprintf("%d", *matchingrule.Id))
+		return nil
+	}
+
+	opts := []datatypes.Network_SecurityGroup_Rule{
+		sgrule,
+	}
+	log.Println("[INFO] creating security group rule")
+	_, err = service.Id(securitygroupid).AddRules(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating Security Group Rule: %s", err)
+	}
+
+	matchingrule, err = findMatchingRule(securitygroupid, &sgrule, service)
+	if err != nil {
+		return err
+	}
+	d.SetId(strconv.Itoa(*matchingrule.Id))
+
+	return resourceIBMSecurityGroupRuleRead(d, meta)
+}
+
+func resourceIBMSecurityGroupRuleRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	matchingrules, err := service.Filter(filter.Build(
+		filter.Path("rules.id").Eq(d.Id()))).Id(
+		d.Get("security_group_id").(int)).GetRules()
+	if err != nil {
+		// If the group is somehow already destroyed, mark as
+		// succesfully gone
+		if err, ok := err.(sl.Error); ok && err.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error retrieving Security Group Rule: %s", err)
+	}
+
+	if len(matchingrules) == 0 {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("direction", matchingrules[0].Direction)
+
+	if matchingrules[0].Ethertype != nil {
+		d.Set("ether_type", matchingrules[0].Ethertype)
+	}
+	if matchingrules[0].PortRangeMin != nil {
+		d.Set("port_range_min", *matchingrules[0].PortRangeMin)
+	}
+	if matchingrules[0].PortRangeMax != nil {
+		d.Set("port_range_max", *matchingrules[0].PortRangeMax)
+	}
+	if matchingrules[0].Protocol != nil {
+		d.Set("protocol", matchingrules[0].Protocol)
+	}
+
+	return nil
+}
+
+func resourceIBMSecurityGroupRuleDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetNetworkSecurityGroupService(sess)
+
+	id, _ := strconv.Atoi(d.Id())
+	ids_to_delete := []int{id}
+	_, err := service.Id(d.Get("security_group_id").(int)).RemoveRules(ids_to_delete)
```

Comment:
If you want to ensure that a security group is deleted in this function, along with removing all rules, you'll also want to detach all attached interfaces from it. A security group with any attached interfaces cannot be deleted.

It can also be more complex, but if this security group is referenced as a remote security group by a rule in any other groups, it also cannot be deleted. The IDs of the groups where they are still referenced will be returned as an error in the delete, so you can use that message to find where to clean up more of these rules.